app/src/main/java/org/dhis2/utils/granularsync/GranularSyncPresenterImpl.kt

```diff
@@ -260,6 +290,58 @@ class GranularSyncPresenterImpl(
         states.postValue(statesList)
     }
 
+    override fun onSmsNotManuallySent(context: Context) {
+        view.logSmsNotSent()
+        smsSyncProvider.unregisterSMSReceiver(context)
+    }
+
+    override fun onSmsSyncClick(
+        callback: (LiveData<List<SmsSendingService.SendingStatus>>) -> Unit
+    ) {
+        if (smsSyncProvider.isPlayServicesEnabled()) {
+            view.logOpeningSmsApp()
+            initSMSSyncPlayServices()
+        } else if (view.checkSmsPermission()) {
+            callback(initSMSSync())
+        }
+    }
+
+    override fun onSmsManuallySent(
+        context: Context,
+        confirmationCallback: (LiveData<Boolean?>) -> Unit
+    ) {
+        if (smsSyncProvider.expectsResponseSMS()) {
+            view.logWaitingForServerResponse()
+            smsSyncProvider.waitForSMSResponse(
+                context,
+                smsSyncProvider.getGatewayNumber(),
+                onSuccess = {
+                    confirmationCallback(smsSyncProvider.observeConfirmationNumber())
+                },
+                onFailure = {
+                    view.logSmsReachedServerError()
+                    updateStatusToSentBySMS()
+                }
+            )
+        } else {
+            view.logSmsSent()
+            updateStatusToSentBySMS()
+        }
+    }
+
+    override fun onConfirmationMessageStateChanged(messageReceived: Boolean?) {
+        when (messageReceived) {
+            true -> {
+                view.logSmsReachedServer()
+                updateStatusToSyncedWithSMS()
+            }
+            false -> {
+                view.logSmsReachedServerError()
+                updateStatusToSentBySMS()
+            }
+        }
+    }
```

Comment:
can we add unit tests?

Perhaps it makes sense to call waitForResponse outside the presenter since it has context dependency